### PR TITLE
feat(as-4052): develop Enforcement Notice

### PR DESCRIPTION
**Pull request by nssdesikan**
## Ticket Number

https://pins-ds.atlassian.net/browse/AS-4052

## Description of change
Develop Enforcement Notice

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `master` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/forms-web-app/src/controllers/before-you-start/enforcement-notice.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+const logger = require('../../lib/logger');
+const { createOrUpdateAppeal } = require('../../lib/appeals-api-wrapper');
+const { VIEW } = require('../../lib/views');
+const {
+  validEnforcementNoticeOptions,
+} = require('../../validators/before-you-start/enforcement-notice');
+
+exports.getServiceNotAvailableWhenReceivedEnforcementNotice = (req, res) => {
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/controllers/before-you-start/enforcement-notice.js:8]: Isn't this functionality included in the use-a-different-notice ticket?

File: packages/forms-web-app/src/routes/before-you-start/enforcement-notice.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+const fetchExistingAppealMiddleware = require('../../middleware/fetch-existing-appeal');
+const enforcementNoticeController = require('../../controllers/before-you-start/enforcement-notice');
+const { validationErrorHandler } = require('../../validators/validation-error-handler');
+const {
+  rules: enforcementNoticeValidationRules,
+} = require('../../validators/before-you-start/enforcement-notice');
+
+const router = express.Router();
+
+router.get(
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/routes/before-you-start/enforcement-notice.js:12]: Isn't this functionality included in the use-a-different-notice ticket?

File: packages/forms-web-app/src/controllers/before-you-start/enforcement-notice.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+exports.getEnforcementNotice = (req, res) => {
+  res.render(VIEW.BEFORE_YOU_START.ENFORCEMENT_NOTICE, {
+    appeal: req.session.appeal,
+  });
+};
+
+exports.postEnforcementNotice = async (req, res) => {
+  const { body } = req;
+  const { errors = {}, errorSummary = [] } = body;
+
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/controllers/before-you-start/enforcement-notice.js:21]: Redundant line space

File: packages/forms-web-app/src/views/before-you-start/enforcement-notice.njk [hunk 4]
```diff
@@ -0,0 +1,74 @@
+{% extends "layouts/main.njk" %}
+
+{% from "govuk/components/date-input/macro.njk" import govukDateInput %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+{% from "govuk/components/error-summary/macro.njk" import govukErrorSummary %}
+{% from "govuk/components/radios/macro.njk" import govukRadios %}
+
+{% set title="Have you received an enforcement notice?- Before you start - Appeal a planning decision - GOV.UK" %}
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/views/before-you-start/enforcement-notice.njk:8]: Should there be a space after notice?

File: packages/forms-web-app/src/validators/before-you-start/enforcement-notice.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+const { body } = require('express-validator');
+
+const validEnforcementNoticeOptions = ['yes', 'no'];
+
+const ruleEnforcementNotice = () =>
+  body('enforcement-notice')
+    .notEmpty()
+    .withMessage('Select Yes if you have received an enforcement notice')
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/validators/before-you-start/enforcement-notice.js:8]: The y in yes should be lowercase, according to the ticket.

File: packages/forms-web-app/src/lib/views.js [hunk 6]
```diff
@@ -55,6 +55,8 @@ const VIEW = {
     ANY_OF_FOLLOWING: 'before-you-start/any-of-following',
     LOCAL_PLANNING_DEPARTMENT: 'before-you-start/local-planning-department',
     TYPE_OF_PLANNING_APPLICATION: 'before-you-start/type-of-planning-application',
+    ENFORCEMENT_NOTICE: 'before-you-start/enforcement-notice',
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/lib/views.js:58]: This is fine, but we have a linitng error (space) on line 45

File: packages/forms-web-app/src/lib/empty-appeal.js [hunk 7]
```diff
@@ -4,6 +4,9 @@ module.exports.APPEAL_DOCUMENT = {
     decisionDate: null,
     submissionDate: null,
     state: 'DRAFT',
+    beforeYourStart: {
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/lib/empty-appeal.js:7]: Typo - beforeYouStart

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/enforcement-notice.test.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+jest.mock('../../../../src/validators/before-you-start/enforcement-notice');
+
+describe('routes/before-you-start/enforcement-notice', () => {
+  beforeEach(() => {
+    // eslint-disable-next-line global-require
+    require('../../../../src/routes/before-you-start/enforcement-notice');
+  });
+
+  it('should define the expected routes', () => {
+    expect(get).toHaveBeenCalledWith(
```
Comment (swdpcomputing) [inline packages/forms-web-app/__tests__/unit/routes/before-you-start/enforcement-notice.test.js:18]: Isn't this functionality included in the use-a-different-notice ticket?

File: packages/forms-web-app/src/routes/before-you-start/enforcement-notice.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  '/use-a-different-service',
+  enforcementNoticeController.getServiceNotAvailableWhenReceivedEnforcementNotice
+);
+router.get(
+  '/enforcement-notice',
```
Comment (swdpcomputing) [inline packages/forms-web-app/src/routes/before-you-start/enforcement-notice.js:17]: I don't think we need to collect the existing appeal for the get

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/enforcement-notice.test.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+      '/use-a-different-service',
+      enforcementNoticeController.getServiceNotAvailableWhenReceivedEnforcementNotice
+    );
+    expect(get).toHaveBeenCalledWith(
+      '/enforcement-notice',
+      [fetchExistingAppealMiddleware],
```
Comment (swdpcomputing) [inline packages/forms-web-app/__tests__/unit/routes/before-you-start/enforcement-notice.test.js:24]: I don't think we need to collect the existing appeal for the get

**Comment by swdpcomputing** [comment]
The backlink isn't working properly and doesn't have any unit tests to prove so.

**Review by swdpcomputing — CHANGES_REQUESTED** [review]

File: packages/forms-web-app/src/controllers/before-you-start/enforcement-notice.js (hunk 1 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/controllers/before-you-start/enforcement-notice.js:8]: Removed

File: packages/forms-web-app/src/routes/before-you-start/enforcement-notice.js (hunk 2 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/routes/before-you-start/enforcement-notice.js:12]: Removed

File: packages/forms-web-app/src/controllers/before-you-start/enforcement-notice.js (hunk 3 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/controllers/before-you-start/enforcement-notice.js:21]: Removed

File: packages/forms-web-app/src/views/before-you-start/enforcement-notice.njk (hunk 4 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/views/before-you-start/enforcement-notice.njk:8]: Eagle eye. Space added.

File: packages/forms-web-app/src/validators/before-you-start/enforcement-notice.js (hunk 5 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/validators/before-you-start/enforcement-notice.js:8]: Corrected

File: packages/forms-web-app/src/lib/views.js (hunk 6 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/lib/views.js:58]: Resolved

File: packages/forms-web-app/src/lib/empty-appeal.js (hunk 7 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/lib/empty-appeal.js:7]: Corrected

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/enforcement-notice.test.js (hunk 8 above)
Comment (nssdesikan) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/enforcement-notice.test.js:18]: Removed

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/enforcement-notice.test.js (hunk 10 above)
Comment (nssdesikan) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/enforcement-notice.test.js:24]: As discussed, leaving it for the moment

File: packages/forms-web-app/src/routes/before-you-start/enforcement-notice.js (hunk 9 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/routes/before-you-start/enforcement-notice.js:17]: As discussed, leaving it for the moment

**Review by swdpcomputing — APPROVED** [review]

**Pull request merged**